cmd/zed/zed.d/statechange-led.sh

```diff
@@ -85,91 +87,70 @@ check_and_set_led()
 		else
 			break
 		fi
-        done
+	done
 }
 
 state_to_val()
 {
 	state="$1"
-	if [ "$state" = "FAULTED" ] || [ "$state" = "DEGRADED" ] || \
-	   [ "$state" = "UNAVAIL" ] ; then
-		echo 1
-	elif [ "$state" = "ONLINE" ] ; then
-		echo 0
-	fi
+	case "$state" in
+		FAULTED|DEGRADED|UNAVAIL)
+			echo 1
+			;;
+		ONLINE)
+			echo 0
+			;;
+	esac
 }
 
-# process_pool ([pool])
+# process_pool (pool)
 #
-# Iterate through a pool (or pools) and set the VDEV's enclosure slot LEDs to
-# the VDEV's state.
+# Iterate through a pool and set the vdevs' enclosure slot LEDs to
+# those vdevs' state.
 #
 # Arguments
-#   pool:	Optional pool name.  If not specified, iterate though all pools.
+#   pool:	Pool name.
 #
 # Return
 #  0 on success, 3 on missing sysfs path
 #
 process_pool()
 {
 	pool="$1"
-	rc=0
-
-	# Lookup all the current LED values and paths in parallel
-	#shellcheck disable=SC2016
-	cmd='echo led_token=$(cat "$VDEV_ENC_SYSFS_PATH/fault"),"$VDEV_ENC_SYSFS_PATH",'
-	out=$($ZPOOL status -vc "$cmd" "$pool" | grep 'led_token=')
-
-	#shellcheck disable=SC2034
-	echo "$out" | while read -r vdev state read write chksum therest; do
-		# Read out current LED value and path
-		tmp=$(echo "$therest" | sed 's/^.*led_token=//g')
-		vdev_enc_sysfs_path=$(echo "$tmp" | awk -F ',' '{print $2}')
-		current_val=$(echo "$tmp" | awk -F ',' '{print $1}')
-
-		if [ "$current_val" != "0" ] ; then
-			current_val=1
-		fi
-
-		if [ -z "$vdev_enc_sysfs_path" ] ; then
-			# Skip anything with no sysfs LED entries
-			continue
-		fi
-
-		if [ ! -e "$vdev_enc_sysfs_path/fault" ] ; then
-			#shellcheck disable=SC2030
-			rc=1
-			zed_log_msg "vdev $vdev '$file/fault' doesn't exist"
-			continue;
-		fi
-
-		val=$(state_to_val "$state")
 
-		if [ "$current_val" = "$val" ] ; then
-			# LED is already set correctly
-			continue;
-		fi
-
-		if ! check_and_set_led "$vdev_enc_sysfs_path/fault" "$val"; then
-			rc=1
-		fi
-
-	done
-
-	#shellcheck disable=SC2031
-	if [ "$rc" = "0" ] ; then
-		return 0
-	else
-		# We didn't see a sysfs entry that we wanted to set
-		return 3
-	fi
+	ZPOOL_SCRIPTS_AS_ROOT=1 $ZPOOL status -c "upath,fault_led" "$pool" |
+	    awk '$6 ~ /^\/dev/ && $7 != "-"' | (
+		rc=0
+		while read -r vdev state _ _ _ upath current_val; do
+			dev="${upath##*/}"
+			sysfs_path="$(readlink -f "/sys/block/$dev/device/enclosure_device")"
+
+			if [ "$current_val" != "0" ] ; then
+				current_val=1
+			fi
+
+			val="$(state_to_val "$state")"
+
+			if [ "$current_val" = "$val" ] || [ -z "$val" ] ; then
+				# LED is already set correctly or we don't care
+				continue
+			fi
+
+			if ! check_and_set_led "$sysfs_path/fault" "$val"; then
+				rc=3
+			fi
+		done
+		exit "$rc"
+	)
```

Comment:
This section assumes a fixed ordering of the fields, but that's not the case.  A bad disk will have some kind of message after the checksum field, like:
```
$ ZPOOL_SCRIPTS_AS_ROOT=1 ZPOOL status -c "upath,fault_led"
...
	    U83     FAULTED      0     0     0  external device fault  /dev/sdcj          2
```

Also, this is going to an issue:
```
 			sysfs_path="$(readlink -f "/sys/block/$dev/device/enclosure_device")" 
```
The enclosure directory name is `enclosure_device${vendor_stuff}`, like `/sys/class/block/sdcj/device/enclosure_device:SLOT 84 6E  /` (that is not a typo - there are two spaces after the "6E" in the directory name).  I've seen others that use different format, but they all begin with `enclosure_device`.

Would you mind using the patch I posted instead?  It will deal with these issues, and it correctly worked for me on the enclosure I tested on.

Reply:
ugh, this is much worse than I thought it would be. Used your patch instead of this